manual/tre.1
.TH "TRE" "1" "2022-06-19" "TRE 0\&.4\&.0" "Tre Manual"

.SH NAME
tre \- tree, improved.

.SH SYNOPSIS
.B tre
[path] [options]

.SH DESCRIPTION
.sp
Display file system content in tree diagram.
.sp
Hidden files and those configured to be ignored by \fBgit\fR will be
(optionally) ignored.
.sp
With correct configuration, each displayed file can have a shell alias created
for it, which opens the file in the default editor or an otherwise specified
command.

.SH OPTIONS
.TP
\fB\-a\fR, \fB\-\-all\fR
Print all files and directories, including hidden ones.

.TP
\fB\-d\fR, \fB\-\-directories\fR
Only list directories in output.

.TP
\fB\-e\fR, \fB\-\-editor\fR [\fIEDITOR\fR]
Create aliases for each displayed result in /tmp/tre_aliases_$USER, and add
a number in front of file name to indicate the alias name. For example, a number
"42" means an shell alias "e42" has been created. Running "e42" will cause the
associated file or directory to be open with $EDITOR, or a command specified
along with this command.

.TP
\fB\-j\fR, \fB\-\-json\fR
Output JSON instead of tree diagram.

.TP
\fB\-l\fR, \fB\-\-limit\fR [\fIDEPTH\fR]
Limit display depth file tree output.

.TP
\fB\-s\fR, \fB\-\-simple\fR
Use normal print despite gitignore settings. '-a' has higher priority.

.TP
\fB\-E\fR, \fB\-\-exclude\fR \fIPATTERN\fR
Exclude paths matching a regex pattern. Repeatable.

.TP
\fB\-e\fR, \fB\-\-color\fR \fIWHEN\fR
When to color the output. Choose from \fIautomatic\fR, \fIalways\fR, or \fInever\fR.
\fIautomatic\fR means when printing to a terminal,
tre will include colors; otherwise it will disable colors.

.TP
\fB\-v\fR, \fB\-\-version\fR
Show version number.

.TP
\fB\-h\fR, \fB\-\-help\fR
Show this help message.

.SH REPORTING BUGS
If you notice any issue, Please file a report at
\fIhttps://github.com/dduan/tre/issues\fR

.SH AUTHOR
Daniel Duan (\fIhttps://duan.ca\fR)



CHANGELOG.md
# main

# 0.4.0

- Add M1 builds to release.
- Support installing from Nix flake.
- Add `-E/--exclude PATTERN` option to exclude paths from results.
- Add `-e/--color WHEN` option to control colors in output.
- The short flag for `--version` is renamed from `-v` to `-V`.
- Provide completion scripts for various shells in scripts/completion.
- Add `-p/--portable` option, which enables portable paths when generating editor
  aliases. Without this flag, editor alias only works in the same working directory
  in which `tre` was last invoked.
- Files untracked by git is now included in output by default.
- Output order is now deterministic.

# 0.3.6

- Fixed a bug when root isn't ".", the paths for sub-directories are wrong.

# 0.3.5

- Fix a bug where intermediary directories from `git ls-files` don't have
  correct path, resulting in wrong colors in output, as well as non-functioning
  editor alias.

# 0.3.4

- Add '--limit/-l' option, which limits the maximum diplay depth of the file
  tree.

# 0.3.3

- Hidden directory as the direct input now reveals its content. Previously the
  input directory is not exempt from the "hiding the hidden directry"
  functionality unless `-a` is used. Implemented by [JaSpa][] at #55.

[JaSpa]: https://github.com/JaSpa

# 0.3.2

- Add '--directotries-only/-d' option, which limits output to only directories.
- Add '--json/-j' option, which makes tre output JSON instead of the tree diagram.
- Unix manual is now included as `manual/tre.1`. It maybe distributed along with
  the executable now.

# 0.3.1

Support LS_COLORS, with a default set of colors.

# 0.3.0

Added editor/program aliasing support for Command Prompt and PowerShell on
Windows.

# 0.2.3

Fixed a issue where single quote in file name prevents editor aliasing

# 0.2.2

Improve stability. Reduce possibility of crashing.

# 0.2.1

Add initial Windows build.

# 0.2.0

**Breaking change**: The `-e` option now takes an optional argument. When the
argument is supplied, it'll be used as the command instead of the `$EDITOR`
environment variable for opening the selected path. In previous setup
instructions, the invocation in shell scripts is `tre -e PATH`. In this release,
that command will cause `tre` to think `PATH` is the editor argument. So user
will need to update their setup to be `tre PATH -e` instead (README has been
updated to reflect this.)

This release is rewritten from scratch in Rust, which brings some nice speed up
and Linux binary distribution. In the near futrue, Windows PowerShell support
could be added as well.



scripts/completion/tre.bash
_tre() {
    local i cur prev opts cmds
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    cmd=""
    opts=""

    for i in ${COMP_WORDS[@]}
    do
        case "${i}" in
            "$1")
                cmd="tre"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        tre)
            opts="-h -V -a -s -d -e -j -l -E -c -p --help --version --all --simple --directories --editor --json --limit --exclude --color --portable <PATH>"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --editor)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -e)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --limit)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -l)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --exclude)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -E)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --color)
                    COMPREPLY=($(compgen -W "automatic always never" -- "${cur}"))
                    return 0
                    ;;
                -c)
                    COMPREPLY=($(compgen -W "automatic always never" -- "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

complete -F _tre -o bashdefault -o default tre
